File: LearningLens2025/legacy_files/focused-ai-grader-service/focusedai-platform/backend/src/main/java/com/focusedai/model/grading/Grade.java
package com.focusedai.model.grading;

import java.util.Map;

public class Grade {
    private String gradeId;
    private String submissionId;
    private double score;
    private double maxScore;
    private double percentage;
    private String letterGrade;
    private String feedback;
    private boolean passed;
    private String gradingStrategy;
    private Map<String, Object> executionDetails;
    private Map<String, Object> analysisDetails;
    private long gradedAt;
    private String gradedBy;
    
    // Constructors
    public Grade() {
        this.gradedAt = System.currentTimeMillis();
    }
    
    public Grade(String submissionId, double score, double maxScore) {
        this();
        this.submissionId = submissionId;
        this.score = score;
        this.maxScore = maxScore;
        this.percentage = (score / maxScore) * 100;
        this.letterGrade = calculateLetterGrade(percentage);
        this.passed = score >= (maxScore * 0.6); // 60% passing threshold
    }
    
    // Helper method to calculate letter grade
    private String calculateLetterGrade(double percentage) {
        if (percentage >= 90) return "A";
        if (percentage >= 80) return "B";
        if (percentage >= 70) return "C";
        if (percentage >= 60) return "D";
        return "F";
    }
    
    // Getters and setters
    public String getGradeId() { return gradeId; }
    public void setGradeId(String gradeId) { this.gradeId = gradeId; }
    
    public String getSubmissionId() { return submissionId; }
    public void setSubmissionId(String submissionId) { this.submissionId = submissionId; }
    
    public double getScore() { return score; }
    public void setScore(double score) { this.score = score; }
    
    public double getMaxScore() { return maxScore; }
    public void setMaxScore(double maxScore) { this.maxScore = maxScore; }
    
    public double getPercentage() { return percentage; }
    public void setPercentage(double percentage) { this.percentage = percentage; }
    
    public String getLetterGrade() { return letterGrade; }
    public void setLetterGrade(String letterGrade) { this.letterGrade = letterGrade; }
    
    public String getFeedback() { return feedback; }
    public void setFeedback(String feedback) { this.feedback = feedback; }
    
    public boolean isPassed() { return passed; }
    public void setPassed(boolean passed) { this.passed = passed; }
    
    public String getGradingStrategy() { return gradingStrategy; }
    public void setGradingStrategy(String gradingStrategy) { this.gradingStrategy = gradingStrategy; }
    
    public Map<String, Object> getExecutionDetails() { return executionDetails; }
    public void setExecutionDetails(Map<String, Object> executionDetails) { this.executionDetails = executionDetails; }
    
    public Map<String, Object> getAnalysisDetails() { return analysisDetails; }
    public void setAnalysisDetails(Map<String, Object> analysisDetails) { this.analysisDetails = analysisDetails; }
    
    public long getGradedAt() { return gradedAt; }
    public void setGradedAt(long gradedAt) { this.gradedAt = gradedAt; }
    
    public String getGradedBy() { return gradedBy; }
    public void setGradedBy(String gradedBy) { this.gradedBy = gradedBy; }
}

File: LearningLens2025/legacy_files/focused-ai-grader-service/focusedai-platform/frontend/lib/core/models/grade.dart
class Grade {
  final String id;
  final String submissionId;
  final double score;
  final double maxScore;
  final double percentage;
  final String letterGrade;
  final String feedback;
  final bool passed;
  final String? strategy;
  final Map<String, dynamic>? executionDetails;
  final DateTime gradedAt;
  final String? gradedBy;

  Grade({
    required this.id,
    required this.submissionId,
    required this.score,
    required this.maxScore,
    required this.percentage,
    required this.letterGrade,
    required this.feedback,
    required this.passed,
    this.strategy,
    this.executionDetails,
    required this.gradedAt,
    this.gradedBy,
  });

  factory Grade.fromJson(Map<String, dynamic> json) {
    return Grade(
      id: json['id'] ?? json['gradeId'] ?? '',
      submissionId: json['submissionId'] ?? '',
      score: (json['score'] ?? 0.0).toDouble(),
      maxScore: (json['maxScore'] ?? 100.0).toDouble(),
      percentage: (json['percentage'] ?? 0.0).toDouble(),
      letterGrade: json['letterGrade'] ?? 'F',
      feedback: json['feedback'] ?? '',
      passed: json['passed'] ?? json['isPassed'] ?? false,
      strategy: json['strategy'] ?? json['gradingStrategy'],
      executionDetails: json['executionDetails'],
      gradedAt: json['gradedAt'] != null
          ? DateTime.parse(json['gradedAt'])
          : DateTime.now(),
      gradedBy: json['gradedBy'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'submissionId': submissionId,
      'score': score,
      'maxScore': maxScore,
      'percentage': percentage,
      'letterGrade': letterGrade,
      'feedback': feedback,
      'passed': passed,
      if (strategy != null) 'strategy': strategy,
      if (executionDetails != null) 'executionDetails': executionDetails,
      'gradedAt': gradedAt.toIso8601String(),
      if (gradedBy != null) 'gradedBy': gradedBy,
    };
  }

  Grade copyWith({
    String? id,
    String? submissionId,
    double? score,
    double? maxScore,
    double? percentage,
    String? letterGrade,
    String? feedback,
    bool? passed,
    String? strategy,
    Map<String, dynamic>? executionDetails,
    DateTime? gradedAt,
    String? gradedBy,
  }) {
    return Grade(
      id: id ?? this.id,
      submissionId: submissionId ?? this.submissionId,
      score: score ?? this.score,
      maxScore: maxScore ?? this.maxScore,
      percentage: percentage ?? this.percentage,
      letterGrade: letterGrade ?? this.letterGrade,
      feedback: feedback ?? this.feedback,
      passed: passed ?? this.passed,
      strategy: strategy ?? this.strategy,
      executionDetails: executionDetails ?? this.executionDetails,
      gradedAt: gradedAt ?? this.gradedAt,
      gradedBy: gradedBy ?? this.gradedBy,
    );
  }
}